### PR TITLE
adapt to k8s environment

### LINKED ISSUE #74: deploy to kubernetes
- adapt internal delivery.yaml to pick latest docker images created with this repo's delivery.yaml

### PR DESCRIPTION
**Pull request by lotharschulz**
fix #74 #55 #76 .

- detects cloud environment (AWS or K8s) based on downward API label file in addition to taupage yaml file check
- introduces a delivery.yaml file to build on Kubernetes / CDP
- uses a statefulset to run ghe-backup on Kubernetes
- only one Dockerfile using docker cache
- dockerfile & cron w/o root

### CODE REVIEW

**Comment by lotharschulz** [comment]
docker build with docker cache in place

**Comment by lotharschulz** [comment]
public key file missing on ghe appliance:

```
$ zkubectl logs -l app=ghe-backup                                           
The file ~/.ssh/id_rsa does not exists. Start writing private ssh key.
Private ssh key file written.
start tailing logs
Starting backup of x.zalan.do in snapshot 20171029T225736
ssh: connect to host x.zalan.do port 122: Connection timed out
Error: ssh connection with 'github.bus.zalan.do:122' failed
Note that your SSH key needs to be setup on x.zalan.do:122 as described in:
* https://enterprise.github.com/help/articles/adding-an-ssh-key-for-shell-access
```

**Comment by lotharschulz** [comment]
ghe-backup working now:

```
r@statefulset-ghe-backup-0:/backup/backup-utils# /start_backup.sh
```

```
$ zkubectl exec -it statefulset-ghe-backup-0 bash
r@statefulset-ghe-backup-0:/backup# ps auxwf                                       
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
r       635  0.1  0.0  21468  3960 ?        Ss   10:09   0:00 bash
r       648  0.0  0.0  37388  3312 ?        R+   10:09   0:00  \_ ps auxwf
r       183  0.0  0.0  21476  4072 ?        Ss   10:01   0:00 bash
r       202  0.0  0.0  21304  3612 ?        S+   10:07   0:00  \_ /bin/bash /start_backup.sh
r       204  0.0  0.0  21408  3912 ?        S+   10:07   0:00      \_ bash /backup/backup-utils/bin/ghe-backup
r       617  0.0  0.0  44944  3868 ?        S+   10:07   0:00          \_ ssh -p 122 -o ControlMaster=auto -o ControlPath="/tmp/.
r        48  0.0  0.0  21468  1344 ?        Ss+  07:41   0:00 bash
r        15  0.0  0.0  21468  1344 ?        Ss+  07:10   0:00 bash
r         1  0.0  0.0  21284  1012 ?        Ss   07:05   0:00 /bin/bash /backup/final-docker-cmd.sh
r        13  0.0  0.0  20584   500 ?        Ss   07:05   0:00 cron
r        14  0.0  0.0   7348   104 ?        S    07:05   0:00 tail -F /var/log/ghe-prod-backup.log
r@statefulset-ghe-backup-0:/backup# ls /data/ghe-production-data/                                                                                
20171029T225736  20171030T074513  20171030T100750  in-progress
r@statefulset-ghe-backup-0:/backup# ls -lisa /data/ghe-production-data/
total 24
12 4 drwxr-xr-x. 5 r r 4096 Oct 30 10:07 .
 2 4 drwxr-xr-x. 4 r r 4096 Oct 29 22:57 ..
13 4 drwxr-xr-x. 2 r r 4096 Oct 29 22:57 20171029T225736
15 4 drwxr-xr-x. 2 r r 4096 Oct 30 07:45 20171030T074513
17 4 drwxr-xr-x. 3 r r 4096 Oct 30 10:07 20171030T100750
19 4 -rw-r--r--. 1 r r   20 Oct 30 10:07 in-progress
root@statefulset-ghe-backup-0:/backup# ls -lisa /data/ghe-production-data/
total 20
12 4 drwxr-xr-x. 5 r r 4096 Oct 30 10:12 .
 2 4 drwxr-xr-x. 4 r r 4096 Oct 29 22:57 ..
13 4 drwxr-xr-x. 2 r r 4096 Oct 29 22:57 20171029T225736
15 4 drwxr-xr-x. 2 r r 4096 Oct 30 07:45 20171030T074513
17 4 drwxr-xr-x. 3 r r 4096 Oct 30 10:11 20171030T100750
```

**Comment by lotharschulz** [comment]
error message: `/backup/final-docker-cmd.sh: line 9: cron: command not found` appeared and causes pod `CrashLoopBackOff` since it is not root user anymore. talked to @jmcs about options we see:
- use root user as before - maybe this should be a seperate PR
- find alternatives to cron @tuxlife : do you you know other options than cron that are easy to call as non root

**Comment by rashamalek** [comment]
In [final-docker-cmd.sh](https://github.com/zalando/ghe-backup/blob/master/final-docker-cmd.sh)  probably using cron's full path (/usr/sbin/cron) will solve [the issue.](https://github.com/zalando/ghe-backup/pull/76#issuecomment-341037206)

File: Dockerfile [hunk 1]
```diff
@@ -1,23 +1,32 @@
 FROM registry.opensource.zalan.do/stups/python:3.5-cd26
 MAINTAINER lothar.schulz@zalando.de
 
-USER root
```
Comment (rashamalek) [inline Dockerfile]: As stated in [dockerfile_best_practices](https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#user) using su or sudo is not recommended, instead, after adding the user you could just use "USER application"

File: Dockerfile [hunk 2]
```diff
@@ -33,31 +42,30 @@ COPY start_backup.sh /start_backup.sh
...
-  mkfifo /var/log/ghe-prod-backup.log
-
+  mkfifo /var/log/ghe-prod-backup.log && \
 # delete_instuck_progress log
-RUN \
   touch /var/log/ghe-delete-instuck-progress.log && \
   chown -R application: /var/log/ghe-delete-instuck-progress.log
 
-CMD ["/backup/final-docker-cmd.sh"]
+CMD ["su", "-", "application", "-c", "/backup/final-docker-cmd.sh"]
```
Comment (rashamalek) [inline Dockerfile]: as mentioned earlier, ' "su", "-", "application", "-c", ' could be removed.

File: Dockerfile [hunk 3]
```diff
@@ -33,31 +42,34 @@ COPY start_backup.sh /start_backup.sh
...
+  chown -R application: /var/log/ghe-prod-backup.log && \
   touch /var/log/ghe-delete-instuck-progress.log && \
   chown -R application: /var/log/ghe-delete-instuck-progress.log
 
-CMD ["/backup/final-docker-cmd.sh"]
+USER application
+
+# https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#user mentions to avoid sudo,
+# however cron as part of the final-docker-cmd.sh has to run as
+CMD ["sudo", "/backup/final-docker-cmd.sh"]
```
Comment (rashamalek) [inline Dockerfile]: if you execute /backup/final-docker-cmd.sh with sudo , then /kms/convert-kms-private-ssh-key.sh will create the ssh-key in /root/.ssh folder, which will be inaccessible for application user, so ssh into ghe instance will fail:  https://github.com/zalando/ghe-backup/pull/76/files#diff-2a885483390ec330ff6c432b37530949.
Just faced this in ghe-hubble.
workaround would either using sudo in final-docker-cmd.sh (`sudo /usr/sbin/cron`), or adapting /kms/convert-kms-private-ssh-key.sh to create the ssh-key in the right place.
If you prefer the second way, owner of ~/.ssh/id_rsa should be also changed to application

Comment (lotharschulz) [reply Dockerfile]: Thank you, I deployed on Friday to test but did not check back the deployment results.

**Comment by lotharschulz** [comment]
```sudo: /usr/sbin/cron: command not found```

**Comment by rashamalek** [comment]
make sure cron is installed https://github.bus.zalan.do/machinery/hubble/blob/master/Dockerfile#L9

**Comment by lotharschulz** [comment]
Hmm, was there before - I assume as part of https://github.com/zalando-stups/docker-python/blob/master/Dockerfile. Still does not hurt to install it.

**Comment by lotharschulz** [comment]
```
$ zkubectl exec -it statefulset-ghe-backup-0 bash
application@statefulset-ghe-backup-0:~$ ls /data/ghe-production-data
20171109T152700
application@statefulset-ghe-backup-0:~$ ls /data/ghe-production-data/20171109T152700/
authorized-keys.json  enterprise.ghl  manage-password+	redis.rdb      settings.json	  ssl-ca-certificates.tar  uuid
benchmarks	      incomplete      mysql.sql.gz	saml-keys.tar  ssh-host-keys.tar  strategy		   version
application@statefulset-ghe-backup-0:~$ ps auxwf
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
applica+  1898  0.0  0.0  21472  4024 ?        Ss   15:37   0:00 bash
applica+  1909  0.0  0.0  37392  3304 ?        R+   15:37   0:00  \_ ps auxwf
applica+    16  0.0  0.0  21560  4060 ?        Ss+  15:09   0:00 bash
applica+     1  0.0  0.0   4536   880 ?        Ss   15:07   0:00 /bin/sh -c "/backup/final-docker-cmd.sh"
applica+     6  0.0  0.0  21300  3588 ?        S    15:07   0:00 /bin/bash /backup/final-docker-cmd.sh
applica+    15  0.0  0.0   7352   772 ?        S    15:07   0:00  \_ tail -F /var/log/ghe-prod-backup.log
root        14  0.0  0.0  29012  2696 ?        Ss   15:07   0:00 /usr/sbin/cron
applica+  1331  1.3  0.0  47204  2872 ?        Ss   15:27   0:08 ssh -f -p 122 -o ControlMaster=auto -o ControlPath="/tmp/.ghe-sshmux-185980bf" -o ControlPersist=10m -o ServerAliv

```

**Comment by lotharschulz** [comment]
screen shot 2017-11-09 at 18 17 16 

need to check the disk size

**Comment by lotharschulz** [comment]
last Friday deployment successfully performed backups according to cron schedule:

```
application@statefulset-ghe-backup-0:~$ cat /etc/cron.d/ghe-backup
# ghe automata prod backups
# timezone is GMT -1
3 22 * * 1-7 application /start_backup.sh
```

```
application@statefulset-ghe-backup-0:~$ ls -lisa /data/ghe-production-data/
total 24
26214401 4 drwxr-xr-x.  6 application application 4096 Nov 12 22:16 .
       2 4 drwxr-xr-x.  4 application application 4096 Nov 10 16:23 ..
26214402 4 drwxrwxr-x. 11 application application 4096 Nov 10 22:13 20171110T220301
26403018 4 drwxrwxr-x. 11 application application 4096 Nov 11 22:18 20171111T220301
26222904 4 drwxrwxr-x. 11 application application 4096 Nov 12 22:16 20171112T220301
26214403 0 lrwxrwxrwx.  1 application application   15 Nov 12 22:16 current -> 20171112T220301
application@statefulset-ghe-backup-0:~$ ls /data/ghe-production-data/20171112T220301       
alambic_assets	      benchmarks      es-scan-complete	manage-password+  redis.rdb	 settings.json		  storage   version
audit-log	      elasticsearch   git-hooks		mysql.sql.gz	  repositories	 ssh-host-keys.tar	  strategy
authorized-keys.json  enterprise.ghl  hookshot		pages		  saml-keys.tar  ssl-ca-certificates.tar  uuid
```

**Comment by lotharschulz** [comment]
restore attempt started with 

`application@statefulset-ghe-backup-0:~/backup-utils/bin$ nohup ./ghe-restore -f [ghe master public IP] &`

**Comment by lotharschulz** [comment]
restore result: restoring in general works, the target instance disk was just to small

```
application@statefulset-ghe-backup-0:~$ cd /backup/backup-utils/bin/
application@statefulset-ghe-backup-0:~/backup-utils/bin$ tail -f nohup.out
WARNING: f/nw/fa/ac/bc/1816/2253.git/audit_log failed verification -- update discarded (will try again).
Restoring GitHub Pages ...
Restoring SSH authorized keys ...
Restoring asset attachments ...
Restoring storage data ...
Restoring hook deliveries ...
Restoring custom Git hooks ...
Restoring Elasticsearch indices ...
rsync: write failed on "/data/user/elasticsearch-restore/github-enterprise/nodes/0/indices/code-search-3/0/index/_3mg4.fdt": No space left on device (28)
rsync error: error in file IO (code 11) at receiver.c(393) [receiver=3.1.1]
```

**Comment by lotharschulz** [comment]
volume increase (200-800GB) initiated

**Comment by lotharschulz** [comment]
prepare yet another backup and restore attempt

**Comment by lotharschulz** [comment]
```
admin@testbed:~$ ghe-storage-extend
Growing physical volume from 200.00g to 800.00g...
  Physical volume "/dev/xvdb" changed
  1 physical volume(s) resized / 0 physical volume(s) not resized

```

**Comment by lotharschulz** [comment]
schedule backup approach tonight:

```
...
33 23 * * 1-6 application /start_backup.sh
...
```

**Comment by lotharschulz** [comment]
- schedule backup approach failed due to a issue that was fixed with https://github.com/zalando/ghe-backup/pull/76/commits/62035f3683af5c8c3f1e7605a7498160d93a793f
- new backup approach scheduled for
`39 10 * * 1-6 application /start_backup.sh`

**Comment by lotharschulz** [comment]
backup started as schedule at in folder 103901: `20171116T103901`

**Comment by lotharschulz** [comment]
restore successfully finished:

```
application@statefulset-ghe-backup-0:~$ tail -f nohup.out 
Restoring asset attachments ...
Restoring storage data ...
Restoring hook deliveries ...
Restoring custom Git hooks ...
Restoring Elasticsearch indices ...
Configuring storage ...
Starting cron ...
Restoring SSH host keys ...
Completed restore of [IP]:122 from snapshot 20171116T103901
Visit https://[IP]/setup/settings to review appliance configuration.
```

Visiting the management console supports that:

screen shot 2017-11-16 at 19 01 58

**Comment by lotharschulz** [comment]
@rashamalek @tuxlife @jmcs : 👍  please

**Comment by lotharschulz** [comment]
👍

**Comment by rashamalek** [comment]
👍

**Comment by lotharschulz** [comment]
[code in master needs to make sure Pier One push happens from master only](https://github.com/zalando/ghe-backup/pull/76/commits/0b7d95d5c6dab72b43503f62e90f98383a514ff9).

**Comment by lotharschulz** [comment]
👍

File: Dockerfile [hunk 4]
```diff
@@ -1,23 +1,36 @@
-FROM registry.opensource.zalan.do/stups/python:3.5-cd26
+FROM registry.opensource.zalan.do/stups/python:3.6.2-14
```
Comment (rashamalek) [inline Dockerfile:1]: latest version of stups/python  is 3.6.3-15 published 42h ago

Comment (lotharschulz) [reply Dockerfile:1]: indeed lets do this with a new ticket: https://github.com/zalando/ghe-backup/issues/78

**Comment by lotharschulz** [comment]
👍

**Comment by rashamalek** [comment]
👍

**Comment by jmcs** [comment]
:+1:

**Pull request merged**